[Op] Add `StringTensorPack-15` and `StringTensorUnpack-15` specifications (#25104)
openvinotoolkit/openvino: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/type/string-tensor-unpack-15.rst

    * ``begins[1]`` is equal to 5, because length of the string "Intel" is equal to 5.
    * ``begins.shape`` is equal to [2], because the ``input`` is a batch of 2 strings.

* *ends* = [5, 13]
    * ``ends[0]`` is equal to 5, because length of the string "Intel" is equal to 5.
    * ``ends[1]`` is equal to 13, because length of the string "OpenVINO" is 8, and it needs to be summed up with length of the string "Intel".
    * ``ends.shape`` is equal to ``[2]``, because the ``input`` is a batch of 2 strings.

* *symbols* = "IntelOpenVINO"
    * ``symbols`` contains concatenated string data encoded in utf-8 bytes, interpretable using ``begins`` and ``ends``.
    * ``symbols.shape`` is equal to ``[13]``, because it's the length of concatenated ``input`` strings.

When defining *begins* and *ends*, the notation ``[a, b)`` is used. This means that the range starts with ``a`` and includes all values up to, 
but not including, ``b``. That is why in the example given the length of "IntelOpenVINO" is 12, but *ends* vector contains 13.

**Inputs**
  
* **1**: ``data`` - ND tensor of type *string*. **Required.** 

**Outputs**

> * **1**: ``begins`` - ND tensor of non-negative integer numbers of type *int32* and of the same shape as ``data``.
> 
> * **2**: ``ends`` - ND tensor of non-negative integer numbers of type *int32* and of the same shape as ``data``.
> 

Reviewer: ```suggestion
* **1**: ``begins`` - ND tensor of non-negative integer numbers of type *int32* and of the same shape as ``data`` input.

* **2**: ``ends`` - ND tensor of non-negative integer numbers of type *int32* and of the same shape as ``data`` input.

```

> * **3**: ``symbols`` - 1D tensor of concatenated strings data encoded in utf-8 bytes, of type *u8* and length ``[1]``.

Reviewer: The info about lenght `[1]` seems to be not relevant for `sybmols` input:
```suggestion
* **3**: ``symbols`` - 1D tensor of concatenated strings data encoded in utf-8 bytes, of type *u8* and size equal to the sum of the lengths of each string from the ``data`` input.
```

Author: Fixed seconds before the review, but I applied your version anyway as it sounds better.

* *begins* = [0, 5]
    * ``begins[0]`` is equal to 0, because the first string starts at the beginning index.
    * ``begins[1]`` is equal to 5, because length of the string "Intel" is equal to 5.
    * ``begins.shape`` is equal to [2], because the ``input`` is a batch of 2 strings.

* *ends* = [5, 13]
    * ``ends[0]`` is equal to 5, because length of the string "Intel" is equal to 5.
    * ``ends[1]`` is equal to 13, because length of the string "OpenVINO" is 8, and it needs to be summed up with length of the string "Intel".
    * ``ends.shape`` is equal to ``[2]``, because the ``input`` is a batch of 2 strings.

* *symbols* = "IntelOpenVINO"
    * ``symbols`` contains concatenated string data encoded in utf-8 bytes, interpretable using ``begins`` and ``ends``.
    * ``symbols.shape`` is equal to ``[13]``, because it's the length of concatenated ``input`` strings.

When defining *begins* and *ends*, the notation ``[a, b)`` is used. This means that the range starts with ``a`` and includes all values up to, 
but not including, ``b``. That is why in the example given the length of "IntelOpenVINO" is 12, but *ends* vector contains 13.

> **Inputs**
> 
> * **1**: *data*
> 
>   * **Description**: A tensor containing a string to be unpacked. **Required.**
>   * **Range of values**: A string tensor.
>   * **Type**: ``string``

Reviewer: Taking into account previous suggestion, the way of Inputs/Outputs description is still not aligned with the rest of the specifications. 
```suggestion
**Inputs**

* **1**: ``data`` - A tensor of type ``string`` and arbitrary ND shape, containing string data to be unpacked. **Required.**
```
Currently it follows attributes format instead, please take as an example any other specification for instance: https://github.com/openvinotoolkit/openvino/blob/9dcccbca5424c622805e1e8f9659b5ee0dcb1c87/docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.rst?plain=1#L53-L72

Author: Fixed, the style is now aligned.

> **Outputs**
> 
> * **1**: *begins*:
> 
>   * **Description**: Indices of each string's beginnings.
>   * **Range of values**: ND tensor of non-negative integer numbers.
>   * **Type**: ``int32``

Reviewer: Same comment about alignment, as for the inputs:

```suggestion
**Outputs**

* **1**: ``begins``  - ND tensor of type ``int32`` and the same shape as the ``data`` input tensor, containing indices (non-negative integer numbers) of each string's beginnings.

```

... for all of the outputs

Author: Done.

* **2**: *ends*:

  * **Description**: Indices of each string's endings.
  * **Range of values**: ND tensor of non-negative integer numbers.
  * **Type**: ``int32``

* **3**: *symbols*:

  * **Description**: Concatenated ``input`` strings.
  * **Range of values**: 1D tensor of concatenated strings data encoded in utf-8 bytes.
  * **Type**: ``u8``

**Examples**

*Example 1: input data as string*

For ``input = ["Intel", "OpenVINO"]``

.. code-block:: xml
   :force:

    <layer ... type="StringTensorUnpack" ... >
        <input>

>             <port id="0" precision="string">
>                 <dim>2</dim>     <!-- batch of strings -->
>             </port>
>         </input>
>         <output>
>             <port id="0" precision="I32">
>                 <dim>2</dim>     <!-- begins = [0, 5] -->
>             </port>
>             <port id="1" precision="I32">
>                 <dim>2</dim>     <!-- ends = [5, 13] -->
>             </port>
>             <port id="2" precision="u8">

Reviewer: Detail: To be consistent with IRs and keep it unified within the example, precision should be serialized as upper case (same comment for other examples).
```suggestion
            <port id="0" precision="STRING">
                <dim>2</dim>     <!-- batch of strings -->
            </port>
        </input>
        <output>
            <port id="0" precision="I32">
                <dim>2</dim>     <!-- begins = [0, 5] -->
            </port>
            <port id="1" precision="I32">
                <dim>2</dim>     <!-- ends = [5, 13] -->
            </port>
            <port id="2" precision="U8">
```

Author: Done.

.. {#openvino_docs_ops_type_StringTensorUnpack_15}

StringTensorUnpack
===================

.. meta::
  :description: Learn about StringTensorUnpack-15 - operation which unpacks a batch of strings into three tensors.

**Versioned name**: *StringTensorUnpack-15*

**Category**: *Type*

**Short description**: *StringTensorUnpack* operation transforms a given batch of strings into three tensors - one containing 
the concatenated string data, and two other storing begin and end indices of the strings, respectively.

**Detailed description**

Consider an ``input`` string tensor containing values ``["Intel", "OpenVINO"]``.

The operator will transform the tensor into three outputs:

* *begins* = [0, 5]

>     * ``begins[0]`` is equal to 0, because the first string starts at the beggining index.

Reviewer 2: ```suggestion
    * ``begins[0]`` is equal to 0, because the first string starts at the beginning index.
```

    * ``begins[1]`` is equal to 5, because length of the string "Intel" is equal to 5.
    * ``begins.shape`` is equal to [2], because the ``input`` is a batch of 2 strings.

* *ends* = [5, 13]
    * ``ends[0]`` is equal to 5, because length of the string "Intel" is equal to 5.

>     * ``ends[1]`` is equal to 13, because length of the string "OpenVINO" is 8, and it needs to be summed up
>     with length of the string "Intel".
>     * ``ends.shape`` is equal to ``[2]``, because the ``input`` is a batch of 2 strings.

Reviewer 2: ```suggestion
    * ``ends[1]`` is equal to 13, because length of the string "OpenVINO" is 8, and it needs to be summed up with length of the string "Intel".
    * ``ends.shape`` is equal to ``[2]``, because the ``input`` is a batch of 2 strings.
```
Last element in list is not properly formatted in generated docs

* *symbols* = "IntelOpenVINO"
    * ``symbols`` contains concatenated string data encoded in utf-8 bytes, interpretable using ``begins`` and ``ends``.
    * ``symbols.shape`` is equal to ``[13]``, because it's the length of concatenated ``input`` strings.

When defining *begins* and *ends*, the notation ``[a, b)`` is used. This means that the range starts with ``a`` and includes all values up to, 
but not including, ``b``. That is why in the example given the length of "IntelOpenVINO" is 12, but *ends* vector contains 13.

**Inputs**

* **1**: *data*

  * **Description**: A tensor containing a string to be unpacked. **Required.**
  * **Range of values**: A string tensor.
  * **Type**: ``string``

**Outputs**

* **1**: *begins*:

>   * **Description**: Indices of each string's begginings.

Reviewer 2: ```suggestion
  * **Description**: Indices of each string's beginnings.
```

  * **Range of values**: ND tensor of non-negative integer numbers.
  * **Type**: *T_IDX*

* **2**: *ends*:

  * **Description**: Indices of each string's endings.
  * **Range of values**: ND tensor of non-negative integer numbers.
  * **Type**: *T_IDX*

* **3**: *symbols*:

  * **Description**: Concatenated ``input`` strings.
  * **Range of values**: 1D tensor of concatenated strings data encoded in utf-8 bytes.
  * **Type**: ``u8``

**Types**

> * *T_IDX*: ``int32``

Reviewer 2: Since T_IDX is single type, maybe remove this section and use i32 directly?

Author: Done.

Reviewer: Defining it as `T_IDX` within the **Types** section is a clear and common approach, even if it's the only type, it's easier to manage. But the updated version is also acceptable.

**Examples**

*Example 1: input data as string*

For ``input = ["Intel", "OpenVINO"]``

.. code-block:: xml
   :force:

    <layer ... type="StringTensorUnpack" ... >
        <input>
            <port id="0" precision="string">
                <dim>2</dim>     <!-- batch of strings -->
            </port>
        </input>
        <output>

>             <port id="0" precision="I64">

Reviewer 2: Spec defines output type as i32. Same to other examples

Author: Fixed.
